raft-tech/TANF-app
# feat(1136): Accept AMS OpenID

### Summary of Changes

This PR adds initial support in the backend for accepting AMS OpenID As an authentication provider.

Closes #1136

_Acceptance criteria as stated in the issue_

### How to Test
_List the steps to test the PR_
These steps are generic, please adjust as necessary.
```
cd tdrs-frontend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d
cd tdrs-backend && docker-compose -f docker-compose.yml -f docker-compose.local.yml up -d 
```

1. Open `http://localhost:3000/` or `https://tdp-frontend-raft.app.cloud.gov` and sign in using the Login With AMS button.
_IMPORTANT NOTE: given limited acf ams config settings, only one deployed site url (the one above) currently works. we'll inquire if this can change but for now, this will NOT work in staging env_
1. Proceed with the login flow in HHS AMS.
1. Ensure that you are redirected to TDRS
1. Ensure you are able to use the app as normal
1. Log out, and get redirected to the main login page with no errors.
1. Test steps should be captured in the demo GIF(s) and/or screenshots below.

> *Demo GIF(s) and screenshots for testing procedure*

### Deliverable 1: Accepted Features

> *Performance Standard(s): At the beginning of each sprint, the Product Owner and development team will collaborate to define a set of user stories to be completed during the sprint. Acceptance criteria for each story will also be defined. The development team will deliver code and functionality to satisfy these user stories.*

> *Acceptable Quality Level: Delivered code meets the acceptance criteria for each user story. Incomplete stories will be assessed and considered for inclusion in the next sprint.*

 + [x] Authentication and Authorization can pass through successfully using AMS.
 + [x] User's request is redirected back to our frontend
_also confirmed that external user route still works_
 + [x] `https://tdp-backend-raft.app.cloud.gov/admin/` for logged out user re-routes to frontend (addressing #1520 behavior)
 + [x] @ADPennington reviewed and tested


As Product Owner, @lfrohlich will decide if ACs are met.

### Deliverable 2: Tested Code

> *Performance Standard(s): Code delivered under the order must have substantial test code coverage. Version-controlled HHS GitHub repository of code that comprises products that will remain in the government domain.*

> *Acceptable Quality Level: Minimum of 90% test coverage of all code. All areas of code are meaningfully tested.*

+ Are all areas of code introduced in this PR meaningfully tested?
+ Are code coverage minimums met?

### Deliverable 3: Properly Styled Code

> *Performance Standard(s): GSA 18F Front- End Guide*

> *Acceptable Quality Level: 0 linting errors and 0 warnings*


### Deliverable 4: Accessible

> *Performance Standard(s): Web Content Accessibility Guidelines 2.1 AA standards*

> *Acceptable Quality Level: 0 errors reported using an automated scanner and 0 errors reported in manual testing*
    + [See the full Accessibility Assessment plan here.](https://github.com/HHS/TANF-app/blob/main/docs/Technical-Documentation/how-government-will-test-a11y.md)
_n/a: this PR will redirect users to ACF AMS for authentication. ACF AMS meets compliance standards._ 

### Deliverable 5: Deployed

> *Performance Standard(s): Code must successfully build and deploy into the staging environment.*

> *Acceptable Quality Level: Successful deployment by assigning a 'Deploy with CircleCI - {env}' label*

_yes, BUT, as noted above, this only works on the raft deployed site, given configuration limitations beyond TDP team's control._
cc: @lfrohlich @abottoms-coder 
### Deliverable 6: Documented

> *Performance Standard(s): Summary of user stories completed every two weeks. All dependencies are listed and the licenses are documented. Major functionality in the software/source code is documented, including system diagram. Individual methods are documented inline in a format that permits the use of tools such as JSDoc. All non-inherited 800-53 system security controls are documented in the Open Control or OSCAL format and HHS Section 508 Product Assessment Template (PAT) are updated as appropriate.*

> *Acceptable Quality Level: Combination of manual review and automated testing, if available*

_https://pypi.org/project/requests-mock/_

### Deliverable 7: Secure

> *Performance Standard(s): Open Web Application Security Project (OWASP) Application Security Verification Standard 3.0*

> *Acceptable Quality Level: Code submitted must be free of medium- and high-level static and dynamic security vulnerabilities*

_no new issues detected_ 

### Deliverable 8: Context

> * Performance Standard(s): Code must be understandable and contextualized for the reviewers possess the knowledge and background necessary for analysis and constructive criticism to take place.

> * Acceptable Quality Level: Code submitted in the pull request has context.*
